### PR TITLE
Use cub::BlockRadixSort to improve medium length sort performance

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79628

In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

### CODE REVIEW

**Comment by peterbell10** [comment]
Here are my detailed benchmark results for all sizes in `range(0, 4097, 4)`, expressed as speedup relative to the old implementation. All results come from an RTX 2060.

Unstable sort varies from 1-3x speedup
image

Stable sort varies from 1-25x speedup
image

File: aten/src/ATen/native/cuda/SortUtils.cuh [hunk 1]
```diff
@@ -151,4 +153,101 @@ bitonicSortKVInPlace(at::cuda::detail::TensorInfo<K, IndexType> keys,
   }
 }
 
+template <int KeyDims, int ValueDims,
+          int block_size, int items_per_thread,
+          typename K, typename V, typename IndexType>
+C10_LAUNCH_BOUNDS_1(block_size)
+__global__ void
+radixSortKVInPlace(at::cuda::detail::TensorInfo<K, IndexType> keys,
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SortUtils.cuh:161]: is this better than DeviceSegmentedRadix sort that's used for some configurations already? It seems like it should be pretty similar.

Comment (peterbell10) [reply aten/src/ATen/native/cuda/SortUtils.cuh:161]: Yes, it is significantly better. `DeviceSegmentedRadixSort` is used by `launch_stable_sort_kernel` and if you look at the stable sort speedup graph you can see `radixSortKVInPlace` is at worst 1.5x faster and at best 25x faster, depending length of the dimension being sorted.

Comment (ngimel) [reply aten/src/ATen/native/cuda/SortUtils.cuh:161]: Hm, interesting, seems like cub should have gotten it right

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**